File: samples/MessageCenterSample/MessageCenterSample/ViewController.m
#import "ViewController.h"
#import "Swrve.h"
#import "SwrveBaseCampaign.h"

@interface ViewController ()

@property (nonatomic, retain) NSTimer *checkWindowTimer;
@property (nonatomic, retain) UIWindow *previousWindow;

@end

@implementation ViewController

@synthesize campaigns;
@synthesize checkWindowTimer;

- (void)viewDidLoad {
    [super viewDidLoad];
    [self refreshDataSource];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self refreshDataSource];
    [self.tableView reloadData];
    // Observe for the new campaigns
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(newSwrveCampaigns)
                                                 name:@"SwrveUserResourcesUpdated"
                                               object:nil];
    
    // The Swrve SDK creates new UIWindows to display content to avoid
    // creating issues for games etc. However, this means that the controllers
    // do not get their callbacks called.
    //
    // This will be fixed by submitting events to the notification center in
    // an upcoming release.
    //
    // To workaround this we can query the current UIWindow active,
    // if the window is different from the previous iteration we can
    // notify the table view that the status of one of the items changed
    // "unread" -> "read" which corresponds to a red -> green background
    //
    self.checkWindowTimer = [NSTimer scheduledTimerWithTimeInterval:1.0
                                                             target:self
                                                           selector:@selector(checkCurrentWindow)
                                                           userInfo:nil
                                                            repeats:YES];
}

- (void) checkCurrentWindow
{
    UIWindow *window = [[UIApplication sharedApplication] keyWindow];
    if (window != self.previousWindow) {
        self.previousWindow = window;
        // An in-app message or conversation might have been displayed.
        // Notify the table view that the state of a campaign might have changed.
        [self.tableView reloadData];
    }
}

- (void) newSwrveCampaigns
{
    [self refreshDataSource];
    [self.tableView reloadData];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    // Remove observe
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [self.checkWindowTimer invalidate];
}

- (void)refreshDataSource {
    // Obtain latest campaigns and order by the campaign dateStart property
    NSSortDescriptor *dateDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"dateStart" ascending:NO];
    NSArray *sortDescriptors = [NSArray arrayWithObject:dateDescriptor];
    self.campaigns = [[[Swrve sharedInstance].talk messageCenterCampaigns] sortedArrayUsingDescriptors:sortDescriptors];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
#pragma unused (tableView)
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
#pragma unused (tableView)
    return self.campaigns.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MessageCenterCell"];
    
    // Configure the cell
    SwrveBaseCampaign* campaign = [self.campaigns objectAtIndex:indexPath.row];
    
    // Campaign subject
    cell.textLabel.text = campaign.subject;
    [cell.textLabel setBackgroundColor:[UIColor clearColor]];
    [cell.detailTextLabel setBackgroundColor:[UIColor clearColor]];
    
    // Format the start date for display
    NSDateFormatter *dformat = [[NSDateFormatter alloc] init];
    [dformat setDateFormat:@"MMMM dd, yyyy (EEEE) HH:mm:ss z Z"];
    
    // Campaign start date, change cell background colour based on seen / unseen status
    switch(campaign.state.status) {
        case SWRVE_CAMPAIGN_STATUS_UNSEEN:
            cell.detailTextLabel.text = [dformat stringFromDate:campaign.dateStart];
            [cell setBackgroundColor:[UIColor colorWithRed:255.0f/255.0f green:0.0f blue:0.0f alpha:0.4]];
            
            break;
        case SWRVE_CAMPAIGN_STATUS_SEEN:
            cell.detailTextLabel.text = [dformat stringFromDate:campaign.dateStart];
            [cell setBackgroundColor:[UIColor colorWithRed:0.0f green:255.0f/255.0f blue:0.0f/255.0f alpha:0.4]];
            break;
        default:
            cell.detailTextLabel.text = @"Other";
            break;
    }
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 80.0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Display the campaign when clicked
    [[Swrve sharedInstance].talk showMessageCenterCampaign:[self.campaigns objectAtIndex:indexPath.row]];
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Remove campaign from the sliding button
        [tableView beginUpdates];
        [[Swrve sharedInstance].talk removeMessageCenterCampaign:[self.campaigns objectAtIndex:indexPath.row]];
        [self refreshDataSource];
        [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationLeft];
        [tableView endUpdates];
    }
}

@end

File: samples/Swift/MessageCenterSample/MessageCenterSample/ViewController.swift
import UIKit
import SwrveSDK

class ViewController: UITableViewController, SwrveMessageDelegate {

    var campaigns : NSArray!

    override func viewDidLoad() {
        super.viewDidLoad()
        refreshDataSource()
        self.tableView.reloadData()
        NotificationCenter.default.addObserver(self, selector: #selector(ViewController.newSwrveCampaigns(_:)), name:NSNotification.Name(rawValue: "SwrveUserResourcesUpdated"), object: nil)

        SwrveSDK.messaging().showMessageDelegate = self;
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self)
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return campaigns.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "MessageCenterCell", for: indexPath)
        let baseCampaign = campaigns[(indexPath as NSIndexPath).row] as? SwrveCampaign
        cell.textLabel?.text = baseCampaign?.subject

        let dformat : DateFormatter = DateFormatter()
        dformat.dateFormat = "MMMM dd, yyyy (EEEE) HH:mm:ss z Z"
        cell.detailTextLabel?.text = dformat.string(from: (baseCampaign?.dateStart)!)

        // Campaign cell background colour based on seen / unseen status
        if let state = baseCampaign?.state.status {
            switch(state) {
            case SWRVE_CAMPAIGN_STATUS_UNSEEN:
                cell.backgroundColor = UIColor.init(red:1, green:0, blue:0, alpha:0.4);
                break;
            case SWRVE_CAMPAIGN_STATUS_SEEN:
                cell.backgroundColor = UIColor.init(red:0, green:1, blue:0, alpha:0.4);
                break;
            default:
                break;
            }
        }

        return cell
    }


    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        SwrveSDK.messaging().showMessageCenter(campaigns[(indexPath as NSIndexPath).row] as? SwrveCampaign)
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }

    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return UITableViewCellEditingStyle.delete
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == UITableViewCellEditingStyle.delete {
            tableView.beginUpdates()
            SwrveSDK.messaging().removeMessageCenter(campaigns[(indexPath as NSIndexPath).row] as? SwrveCampaign)
            tableView.deleteRows(at: [indexPath], with: UITableViewRowAnimation.left)
            tableView.endUpdates()
        }
    }

    func refreshDataSource() {
        let unsortedCampaigns: NSArray = SwrveSDK.messaging().messageCenterCampaigns() as NSArray
        let descriptor: NSSortDescriptor = NSSortDescriptor(key: "dateStart", ascending: true)
        campaigns = unsortedCampaigns.sortedArray(using: [descriptor]) as NSArray
    }

    func newSwrveCampaigns(_ notification: Notification) {
        refreshDataSource()
        self.tableView.reloadData()
    }

    override var prefersStatusBarHidden : Bool {
        return true
    }

    func messageWillBeHidden(_ viewController: UIViewController)  {
        // An in-app message or conversation will be hidden.
        // Notify the table view that the state of a campaign might have changed.
        self.tableView.reloadData()
    }
}
